### PR TITLE
Update Cache service to handle refresh + add few tests

### PR DESCRIPTION
**Pull request by prajwal-pai77**
This PR majorly updates the Cache service and tree nodes with the following:

- Updates the cache service to handle refresh of cache when required.
- Update the tree nodes to refresh each time user clicks on the node(previously used to fetch from vs code builtin cache).
- Basic tests for refresh cache.
- Fixes the mongoDB Migrate URI bug.

### CODE REVIEW

File: src/util/cacheService/cacheService.ts [hunk 1]
```diff
@@ -106,61 +121,327 @@ export class CacheService {
...
+                        await this.refreshCollectionIndexCache(connection, collection.bucketName, collection.scopeName, collection.name);
+                    }
+                }
+            }
+        }
+    }
+
+
+    // Checks and refreshes the cache for timeout.If the cache is loaded successfully, it refreshes the cache by updating bucket and collection cache if required
+    public checkCacheTimeout = async (bucketTimeOut: number, collectionTimeout: number, forceRefresh: boolean) => {
```
Comment (lokesh-couchbase) [inline src/util/cacheService/cacheService.ts:244]: can we name this refreshCacheOnTimeout?

File: src/extension.ts [hunk 2]
```diff
@@ -200,12 +200,6 @@ export function activate(context: vscode.ExtensionContext) {
     })
   );
 
-  subscriptions.push(
-    vscode.window.registerTreeDataProvider(
-      "couchbase",
-      clusterConnectionTreeProvider
-    )
-  );
```
Comment (lokesh-couchbase) [inline src/extension.ts:208]: We don't need this anymore?
I am confused

File: src/model/IndexDirectory.ts [hunk 3]
```diff
@@ -57,8 +59,12 @@ export class IndexDirectory implements INode {
         try {
             const connection = getActiveConnection();
             //TODO: Change it to not include IndexNode with undefined scope once the issues with undefined scope and collections fixed
+            if (!connection){
+                return [];
+            }
```
Comment (lokesh-couchbase) [inline src/model/IndexDirectory.ts:64]: Did you reach any scenario which led to adding this?

File: src/util/cacheService/cacheService.ts [hunk 4]
```diff
@@ -106,61 +121,327 @@ export class CacheService {
...
+                await this.fullCache(true);
+            }
+        }
+    }
+
+
+    // Updates the Index cache for a particular collection if required
+    public updateCollectionIndexCache = async (connection: IConnection, bucketName: string, scopeName: string, collectionName: string, indexesResult?: QueryIndex[]) => {
+        if (this.cacheStatus === false) {
+            return undefined;
```
Comment (lokesh-couchbase) [inline src/util/cacheService/cacheService.ts:354]: Shouldn't we do a hard cache refresh if it's not cached? Am I missing something?

File: src/util/cacheService/cacheService.ts [hunk 5]
```diff
@@ -106,61 +121,327 @@ export class CacheService {
...
+            }
+        }
+
+
+
+    }
+
+
+    // Refreshes the index and schema cache for a single collection if necessary
+    private refreshCollectionSchemaCache = async (connection: IConnection, bucketName: string, scopeName: string, collectionName: string, collectionTimeout: number, forceRefresh: boolean, result?: QueryResult<any>) => {
```
Comment (lokesh-couchbase) [inline src/util/cacheService/cacheService.ts:265]: use consistent naming of result as queryResult in all functions

File: src/util/util.ts [hunk 6]
```diff
@@ -37,8 +37,9 @@ class LocalState {
     return;
   }
   update(key: string, value: any): Thenable<void> {
-    return new Promise(() => {
+    return new Promise((resolve) => {
       this.state[key] = value;
+      resolve();
```
Comment (lokesh-couchbase) [inline src/util/util.ts:42]: Was it not auto resolving earlier?

File: package.json [hunk 7]
```diff
@@ -37,6 +37,7 @@
     "watch:extension": "tsc -watch -p ./",
     "watch:views": "webpack --watch --mode development",
     "test": "node ./out/test/runTest.js",
+    "testCache": "jest",
```
Comment (lokesh-couchbase) [inline package.json:40]: add it to test as well, so that it can be tested with every ci build

File: src/util/cacheService/cacheService.ts [hunk 8]
```diff
@@ -106,61 +121,327 @@ export class CacheService {
...
+    };
+
+
+
+
+    //Refreshes the cache for all buckets if they are older than timeout or force refresh or not present in the cache
+    private refreshBucketCache = async (connection: IConnection, bucketTimeOut: number, forceRefresh: boolean) => {
+        try {
+            const buckets = await connection.cluster?.buckets().getAllBuckets();
+
```
Comment (lokesh-couchbase) [inline src/util/cacheService/cacheService.ts:186]: Please remove extra lines

File: src/util/cacheService/cacheService.ts [hunk 9]
```diff
@@ -106,61 +121,327 @@ export class CacheService {
...
-            });
+            }
+        } catch (error) {
+            logger.error(`Error while caching bucket '${bucketName}': ${error}`);
+        }
+    };
+
+
+
+
```
Comment (lokesh-couchbase) [inline src/util/cacheService/cacheService.ts:181]: Remove extra white lines

File: src/util/constants.ts [hunk 10]
```diff
@@ -84,6 +84,8 @@ export class Constants {
   public static SWAP_USED = "Swap Used";
   public static CORES_AVAILABLE = "Cores Available";
   public static CPU_STOLE_RATE = "CPU Stole Rate";
+  public static BUCKET_CACHE_EXPIRY_DURATION = 10;
+  public static COLLECTION_CACHE_EXPIRY_DURATION = 10;
```
Comment (lokesh-couchbase) [inline src/util/constants.ts:88]: Will this be increased before release?

**Review by lokesh-couchbase — COMMENTED** [review]
Nice Work!!

File: src/util/constants.ts (hunk 10 above)
Comment (prajwal-pai77) [reply src/util/constants.ts]: Updated

**Review by lokesh-couchbase — APPROVED** [review]
Awesome Work, Ready for pre-release

**Pull request merged**